pike/test/querydirectory.py

```diff
@@ -124,3 +125,277 @@ def test_restart_scan(self):
+            pass            # ensure the test root dir is created
+        # Creates files within the root directory.
+        remaining_files = self.n_entries
+        while remaining_files > 0:
+            batch_futures = []
+            batch_size = 30
+            if remaining_files < batch_size:
+                batch_size = remaining_files
+            for ix in xrange(batch_size):
+                filename = self.filename_pattern.format(remaining_files)
+                path = "{0}\\{1}".format(self.root_dir_name, filename)
+                batch_futures.append((filename, chan.create(tree, path)))
+                remaining_files -= 1
+            for filename, fu in batch_futures:
+                fh = fu.result()
+                chan.close(fh)
+                self.filenames.append(filename)
+        chan.logoff()
+        QueryDirectoryTestMaxMtu.dataset_created = True
+
+    def setUp(self):
+        self.create_dataset()
+
+    @contextmanager
+    def get_test_root(self, chan, tree):
+        root_handle = chan.create(
+            tree,
+            self.root_dir_name,
+            access=pike.smb2.GENERIC_READ,
+            options=pike.smb2.FILE_DIRECTORY_FILE,
+            share=pike.smb2.FILE_SHARE_READ |\
+                  pike.smb2.FILE_SHARE_WRITE |\
+                  pike.smb2.FILE_SHARE_DELETE).result()
+        try:
+            yield root_handle
+        finally:
+            chan.close(root_handle)
+
+    # Enumerate directory at FILE_DIRECTORY_INFORMATION level using the maximum transfer size.
+    def gen_file_directory_info_max_mtu(self, dialect=None, caps=0):
+        """
+        Enumerate directory at FILE_DIRECTORY_INFORMATION level using the maximum transfer size.
+        """
+
+        with self.tree_connect_with_dialect_and_caps(dialect, caps) as (chan, tree):
+            # Retrieves the maximum transaction size to determine how big the command's payload can be (MTU).
+            max_trans_size = chan.connection.negotiate_response.max_transact_size
+
+            # Enumerates the root directory and validates that the expected files are present in it.
+            with self.get_test_root(chan, tree) as root_handle:
+                dir_query = chan.query_directory(
+                    root_handle,
+                    file_information_class=pike.smb2.FILE_DIRECTORY_INFORMATION,
+                    flags=0,
+                    file_index=0,
+                    file_name='{0}*'.format(self.filename_prefix),
+                    output_buffer_length=self.payload_size)
+
+                transaction_size = dir_query[-1].end - dir_query[0].start
+                self.info("Transaction size for query: {0}/{1}".format(
+                                transaction_size, max_trans_size))
+                self.info("Dir entries returned: {0}/{1}".format(
+                                len(dir_query), self.n_entries))
+                self.assertLessEqual(transaction_size,
```

Comment:
On that same note... if the goal is to test that we can do a full size transfer, then this assert doesn't really tell us much.  Guess there's no way to make a set of directory entries match up in size exactly to max_trans_size -- so that's not an option.   But is there some way to beef up this test?  Maybe ensure directory contents are a bit more than will fit in one query_directory request, then do a 2nd query ... and verify that the transaction_size of the 1st query is > max_trans_size - sizeof(1st entry returned by 2nd query).... Um, er, or something like that! :-}  (Okay, maybe it's not worth that much work...)

Reply:
this is a regression test to ensure that the server doesn't overfill the buffer (internally or on the wire). i could probably get a set of dir entries which would fill the max trans size exactly to the byte and that might be worthwhile to pursue. like you mentioned though, that's quite a bit more work for a questionable amount of gain. i'll see what i can do here to make this a bit more edge-y

Reply:
so i did some more experimentation here, and it seems like the closest i can get is 65534 bytes when i use 512 files with 31 character names which theoretically should yield 65536 bytes, but i can't figure out where the missing 2 bytes are, because when the test computes the buffer size returned from the server it counts 65534 bytes.

the other problem is creating 512 files takes a lot longer than creating 34 files with longer names.

i think i'll probably just go with long name files, do two queries like you're suggesting